### PR TITLE
Circuit Query Hash Optimization

### PR DESCRIPTION
**Pull request by Kolezhniuk**

### CODE REVIEW

**Comment by Kolezhniuk** [comment]
- **credentialAtomicQuerySigV2OnChain.circom**
    template instances: 534
    non-linear constraints: 91998
- **credentialAtomicQueryMTPV2OnChain.circom**
    template instances: 533
    non-linear constraints: 75413

**Review by OBrezhniev — APPROVED** [review]

**Review by OBrezhniev — CHANGES_REQUESTED** [review]
Please fix calculation of iterationCount

**Comment by Kolezhniuk** [comment]
> Please fix calculation of iterationCount

fixed

**Review by vmidyllic — APPROVED** [review]

**Review by OBrezhniev — APPROVED** [review]

File: circuits/lib/utils/valueHasher.circom [hunk 1]
```diff
@@ -0,0 +1,59 @@
+pragma circom 2.0.0;
+
+include "../../../node_modules/circomlib/circuits/poseidon.circom";
+
+// Because of the way the Poseidon hash function is implemented for Solidity, the max number of inputs must be 6
+template ValueHasher(valueArraySize) {
```
Comment (AndriianChestnykh) [inline circuits/lib/utils/valueHasher.circom:6]: May be `spongeHash` and `arraySize`? Too specific naming for the util template

**Review by AndriianChestnykh — CHANGES_REQUESTED** [review]

File: circuits/lib/utils/spongeHash.circom [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+include "../../../node_modules/circomlib/circuits/poseidon.circom";
+
+// Because of the way the Poseidon hash function is implemented for Solidity, the max number of inputs must be 6
+template SpongeHash(arraySize) {
+	signal input in[arraySize];
+	signal output out;
+    // batch size is 5 because 1 input is reserved for the hash of the previous iteration
+    var batchSize = 5;
+    var iterationCount = 0;
+	var hashFnBatchSize = 6;
```
Comment (AndriianChestnykh) [inline circuits/lib/utils/spongeHash.circom:12]: Something with the tabulation

File: circuits/lib/utils/valueHasher.circom (hunk 1 above)
Comment (Kolezhniuk) [reply circuits/lib/utils/valueHasher.circom]: renamed

File: circuits/lib/utils/spongeHash.circom [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    for(var i = 0; i < hashFnBatchSize; i++) {
+        firstPoseidon.inputs[i] <== getArrayValueByIndex(in, arraySize, i);
+    }
+
+    var restLength = arraySize - hashFnBatchSize;
+	if (restLength > batchSize) {
+		var r = restLength % batchSize;
+		var diff = r == 0 ? 0 : batchSize - r;
+		iterationCount = (restLength + diff) / batchSize;
+	}
```
Comment (AndriianChestnykh) [inline circuits/lib/utils/spongeHash.circom:23]: Something with the tabulation

File: circuits/credentialAtomicQueryMTPV2OnChain.circom [hunk 4]
```diff
@@ -0,0 +1,20 @@
+pragma circom 2.0.0;
+
+include "./onchain/credentialAtomicQueryMTPOnChain.circom";
+
+/*
+ public signals:
+ userID - user profile id
+ merklized - `1` if claim is merklized
+*/
```
[the remark this answers is not in the document]
Comment (Kolezhniuk) [reply circuits/credentialAtomicQueryMTPV2OnChain.circom]: Removed comments

**Review by AndriianChestnykh — APPROVED** [review]

File: circuits/lib/utils/spongeHash.circom (hunk 3 above)
Comment (Kolezhniuk) [reply circuits/lib/utils/spongeHash.circom:23]: everything is fine in my editor
image

**Pull request merged**